### PR TITLE
[feature] – Audit log functions for firestore collections

### PR DESCRIPTION
**Pull request by andrashee**
First part of a few PRs to keep track of changes to our firebase data.

These triggers are taking care of keeping a history of the data in `{collectionName}-history` table. 
Some examples
- an update to `recipients/aRecipient` will copy the previous value to `recipients-history/aRecipient/history/{currentDate}`. A `status=updated` is added. At the same time, it will update the `last_updated_at` property in `recipients/aRecipient` to now.

- a deletion `recipients/aRecipient` will also copy the previous value to `recipients-history/aRecipient/history/{currentDate}` with `status=deleted`

-   an update to `recipients/aRecipient/payments/xyz` will copy the previous value to `recipients-history/aRecipient/payments/xyz/history/{currentDate}`. A `status=updated` is added. At the same time, it will update the `last_updated_at` property in `recipients/aRecipient/payments/xyzt` to now.
- deletion works the same

Unfortunately, in the triggers, one can't access the auth context and therefore one doesn't know who triggered the change.

I will propose a workaround in other PRs for this.

### CODE REVIEW

File: functions/src/etl/FirestoreAuditor.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import equal from 'deep-equal';
```
Comment (mkue) [inline functions/src/etl/FirestoreAuditor.ts:1]: Could we use the `isEqual()` function from Lodash? I think it does the same thing and then we don't need an additional library...

File: functions/src/etl/FirestoreAuditor.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	auditCollectionTrigger = functions.firestore
+		.document('{collectionId}/{docId}')
+		.onWrite(async (change: Change<DocumentSnapshot>, context: EventContext) => {
+			return this.auditFirestore(change, context);
+		});
+
+	/**
+	 * Triggers changes to documents of sub collections
+	 */
+	auditSubcollectionTrigger = functions.firestore
```
Comment (mkue) [inline functions/src/etl/FirestoreAuditor.ts:32]: Is this needed?

Couldn't we have a single trigger like this one:

```ts
auditCollectionTrigger = functions.firestore
		.document('{collectionId}/{document=**}')
		.onWrite(async (change: Change<DocumentSnapshot>, context: EventContext) => {
			return this.auditFirestore(change, context);
		});
```

This should imo work, see: https://firebase.google.com/docs/firestore/security/rules-structure#recursive_wildcards

Comment (mkue) [reply functions/src/etl/FirestoreAuditor.ts:32]: Then the case of an update in a nested subcollection would also be covered.

**Review by mkue — APPROVED** [review]
Super nice! Nicely tested👌 

If we can avoid an additional library for the equality check, I would prefer that.

File: functions/src/etl/FirestoreAuditor.ts (hunk 2 above)
Comment (andrashee) [reply functions/src/etl/FirestoreAuditor.ts:32]: Ah nice, that would be much better. Let me try that 👍

File: functions/src/etl/FirestoreAuditor.ts (hunk 1 above)
Comment (andrashee) [reply functions/src/etl/FirestoreAuditor.ts:1]: True, much better

**Review by ssandino — APPROVED** [review]

File: functions/src/etl/FirestoreAuditor.ts (hunk 2 above)
Comment (andrashee) [reply functions/src/etl/FirestoreAuditor.ts:32]: Works like a charm 👍

**Pull request merged**